relay-spans/src/span.rs

```diff
@@ -0,0 +1,459 @@
+    /// Attribute keys MUST be unique (it is not allowed to have more than one
+    /// attribute with the same key).
+    #[serde(default)]
+    pub attributes: Vec<KeyValue>,
+    /// dropped_attributes_count is the number of attributes that were discarded. Attributes
+    /// can be discarded because their keys are too long or because there are too many
+    /// attributes. If this value is 0, then no attributes were dropped.
+    #[serde(default)]
+    pub dropped_attributes_count: u32,
+    /// events is a collection of Event items.``
+    #[serde(default)]
+    pub events: Vec<Event>,
+    /// dropped_events_count is the number of dropped events. If the value is 0, then no
+    /// events were dropped.
+    #[serde(default)]
+    pub dropped_events_count: u32,
+    /// links is a collection of Links, which are references from this span to a span
+    /// in the same or different trace.
+    #[serde(default)]
+    pub links: Vec<Link>,
+    /// links is a collection of Links, which are references from this span to a span
+    /// in the same or different trace.
+    #[serde(default)]
+    pub dropped_links_count: u32,
+    /// An optional final status for this span. Semantically when Status isn't set, it means
+    /// span's status code is unset, i.e. assume STATUS_CODE_UNSET (code = 0).
+    #[serde(default)]
+    pub status: Status,
+}
+
+impl OtelSpan {
+    /// sentry_status() returns a status as defined by Sentry based on the span status.
+    pub fn sentry_status(&self) -> &'static str {
+        let status_code = self.status.code.clone();
+
+        if status_code == StatusCode::Unset || status_code == StatusCode::Ok {
+            return "ok";
+        }
+
+        if let Some(code) = self
+            .attributes
+            .clone()
+            .into_iter()
+            .find(|a| a.key == "http.status_code")
+        {
+            if let Some(code_value) = code.value.to_i64() {
+                if let Some(sentry_status) = status_codes::HTTP.get(&code_value) {
+                    return sentry_status;
+                }
+            }
+        }
+
+        if let Some(code) = self
+            .attributes
+            .clone()
+            .into_iter()
+            .find(|a| a.key == "rpc.grpc.status_code")
+        {
+            if let Some(code_value) = code.value.to_i64() {
+                if let Some(sentry_status) = status_codes::GRPC.get(&code_value) {
+                    return sentry_status;
+                }
+            }
+        }
+
+        "unknown_error"
+    }
+}
+
+impl From<OtelSpan> for EventSpan {
+    fn from(from: OtelSpan) -> Self {
+        let mut exclusive_time_ms = 0f64;
+        let mut attributes: Object<Value> = Object::new();
+        let start_timestamp = Utc.timestamp_nanos(from.start_time_unix_nano);
+        let end_timestamp = Utc.timestamp_nanos(from.end_time_unix_nano);
+        for attribute in from.attributes.clone() {
```

Comment:
To prevent unnecessary clones, deconstruct and then use the fields directly:

```suggestion
        let OtelSpan {
            trace_id,
            span_id,
            parent_span_id,
            name,
            attributes: from_attributes,
            ..
        } = from;

        for attribute in from_attributes.clone() {
```

Reply:
I tried to push that as much as I could, I still needed some `clone()` for attributes, let me know if I actually made the situation completely better or just changed it.

Reply:
👍 I pushed a commit to reduce clones even further.